src/utils/viewer/viewer.cpp

```diff
@@ -359,7 +368,46 @@ void Viewer::drawEvent() {
   int DEFAULT_SCENE = 0;
   int sceneID = sceneID_[DEFAULT_SCENE];
   auto& sceneGraph = sceneManager_.getSceneGraph(sceneID);
-  renderCamera_->getMagnumCamera().draw(sceneGraph.getDrawables());
+
+  // XXX
+  auto& drawables = sceneGraph.getDrawables();
+  auto total = drawables.size();
+  size_t visibles = 0;
+  auto& cam = renderCamera_->getMagnumCamera();
+  if (enableFrustumCulling_) {
+    Magnum::SceneGraph::DrawableGroup3D visibleDrawables;
+
+    for (int iDrawable = total - 1; iDrawable >= 0; iDrawable--) {
+      // This, of course, will cause segfault.
+      // for (int iDrawable = 0; iDrawable < total; ++iDrawable)
+      scene::SceneNode& node =
+          dynamic_cast<scene::SceneNode&>(drawables[iDrawable].object());
+      // mvp = proj * modelview
+      auto mvp = cam.projectionMatrix() * cam.cameraMatrix() *
+                 node.absoluteTransformation();
```

Comment:
I think what would be more performant (and simpler in code) is operating on the result of [cam.drawableTransformations()](https://doc.magnum.graphics/magnum/classMagnum_1_1SceneGraph_1_1Camera.html#a9a17bb008136d9e03687f815a189cda2), filtering that, and feeding that to the other [cam.draw()](https://doc.magnum.graphics/magnum/classMagnum_1_1SceneGraph_1_1Camera.html#a09546e57e1fc178f76cb49adb4b5a37e) overload -- you wouldn't need to add/remove drawables from the list and you'd get the transformations calculated as well.

The way it is now, the loop is doing quite a lot of redundant calculations (especially in deep object hierarchies, as `absoluteTransformation()` always goes all the way up the hierarchy with no chance of reusing what was already calculated before) and then `draw()` does that all again to calculate matrices passed to shaders (again with no chance of reusing what you already had calculated here).

Reply:
I read your doc: (https://doc.magnum.graphics/magnum/classMagnum_1_1SceneGraph_1_1Drawable.html#SceneGraph-Drawable-draw-order)
- I did not get it so far how can we "draw just a subset" by using drawableTransformations(), std::sort() and cam.draw() overload?

- I read the magnum source code of "drawableTransformations()". It will apply the cameraMatrix. Basically it computes cameraMatrix * modelAbsTransformMatrix, which transforms the model to the camera space. Then again, the AABB is an OBB now. How can we do the frustum culling at this point?

Reply:
I think the [doc](https://doc.magnum.graphics/magnum/classMagnum_1_1SceneGraph_1_1Camera.html#a9a17bb008136d9e03687f815a189cda2) is a bit vague on function: drawableTransformations(...)
The description simply says: "Drawable transformations". 
I think a better version would be "Drawable transformations in camera space" (so it implies the model transformation matrix has been left-multiplied by cameraMatrix.)